### ISSUE TITLE
Building pyinstaller on manylinux, libcrypt.so.2 not being shared

### ISSUE THREAD

**Issue by dreambeyondorange** [pyinstaller/pyinstaller#7435]
Hi,

We have been developing a distribution based on pyinstaller for our application https://github.com/aws/aws-parallelcluster.  I have been working with a partner team also using pyinstaller but keep running into issues with using manylinux and was hoping to get some insight.  It starts with the following Dockerfile that we borrowed from another team that is building their app with pyinstaller.

```
FROM quay.io/pypa/manylinux1_x86_64

ARG PYTHON_VERSION
# Open SSL is required to install python
ENV OPENSSL_VERSION="1.1.1s"
# Perl is required to install Open SSL
ENV PERL_VERSION="5.10.1"
ENV PYTHON_PREFIX="/usr/local/cpython-exe-builder"
ENV LD_LIBRARY_PATH=$PYTHON_PREFIX/lib:$LD_LIBRARY_PATH
ENV PYINSTALLER_BUILD_DEPS="git libffi-devel zlib-devel bzip2-devel expat-devel ncurses-devel readline-devel tk-devel gdbm-devel db4-devel libpcap-devel xz-devel zip unzip"
ENV OPENSSL_PREFIX="/usr/local/ssl"
ENV PERL_PREFIX="/usr/local/perl-build"
ENV PYINSTALLER_VIRT_ENV=installer-env
ENV PATH=$PERL_PREFIX/bin:$PATH

RUN yum install -y $PYINSTALLER_BUILD_DEPS

RUN cd /tmp \
&& curl -fsSLO "https://www.cpan.org/src/5.0/perl-$PERL_VERSION.tar.gz" \
&& tar -xzf "perl-$PERL_VERSION.tar.gz" \
&& cd perl-$PERL_VERSION \
&& ./Configure -des -Dprefix=$PERL_PREFIX > /dev/null \
&& make > /dev/null \
&& make install > /dev/null

RUN cd /tmp \
&& curl -fsSLO "https://www.openssl.org/source/openssl-$OPENSSL_VERSION.tar.gz" \
&& tar -xzf "openssl-$OPENSSL_VERSION.tar.gz" \
&& cd openssl-$OPENSSL_VERSION \
&& ./config no-ssl2 no-shared -fPIC --prefix=$OPENSSL_PREFIX > /dev/null \
&& make > /dev/null \
&& make install_sw > /dev/null

RUN cd /tmp \
&& curl -fsSLO "https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tgz" \
&& tar -xzf "Python-$PYTHON_VERSION.tgz" \
&& cd Python-$PYTHON_VERSION \
&& ./configure --prefix=$PYTHON_PREFIX --enable-shared --with-openssl=$OPENSSL_PREFIX > /dev/null \
&& make > /dev/null \
&& make install > /dev/null
```
I've tried building python with pyenv and I needed to install openssl in order to get the ssl module, not sure if any of this is related to that though.  No matter how I build python3, it seems to be linked to several shared libraries

```
[root@be39f4e2d867 /]# ldd /usr/local/cpython-exe-builder/bin/python3
        linux-vdso.so.1 =>  (0x00007ffd873e2000)
        libcrypt.so.2 => /usr/local/lib/libcrypt.so.2 (0x00007f18309a3000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1830787000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f1830583000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f1830380000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f1830177000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f182fef4000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f182fb9b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1830bdb000)
```
But these shared libraries are not being bundled with the app and when I run the app on another system I get the missing lib error

```
[26068] Error loading Python lib '/tmp/_MEI7zeOtg/libpython3.9.so.1.0': dlopen: libcrypt.so.2: cannot open shared object file: No such file or directory
```

Here's the spec file 
```
# -*- mode: python ; coding: utf-8 -*-
from PyInstaller.utils.hooks import collect_data_files
from PyInstaller.utils.hooks import collect_all

datas = []
binaries = []
hiddenimports = []
datas += collect_data_files('connexion')
datas += collect_data_files('jsii')
datas += collect_data_files('aws_cdk')
datas += collect_data_files('constructs')
tmp_ret = collect_all('pcluster')
datas += tmp_ret[0]; binaries += tmp_ret[1]; hiddenimports += tmp_ret[2]


block_cipher = None


a = Analysis(
    ['pcluster/cli/src/pcluster/cli/entrypoint.py'],
    pathex=[],
    binaries=binaries,
    datas=datas,
    hiddenimports=hiddenimports,
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
filtered_binaries = a.binaries - TOC([('lib-dynload/readline.cpython-37m-x86_64-linux-gnu.so', None, None)])
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    filtered_binaries,
    a.zipfiles,
    a.datas,
    [],
    name='pc-exe',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=False,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
```

I'd like to be able to use manylinux since we will only need to build once for our linux platforms, but I'm not sure how to deal with this shared library linkage issue.  Any help is greatly appreciated!

I did a bit more testing and if I add the library to the `binaries` on the Analysis, the app succeeds, but I'd rather not be bundling libcrypt.  I guess this is more of a python build question than pyinstaller but I'm not exactly sure where to ask about that.

**Comment by rokm** [comment]
> I did a bit more testing and if I add the library to the `binaries` on the Analysis, the app succeeds, but I'd rather not be bundling libcrypt. 

It's either bundling `libcrypt.so.2` with the application, or requiring the user to install the package that provides it (if it is even available for their distribution). 

For a bit of background on how `libcrypt.so.2` came to be, see [here](https://discuss.python.org/t/pep-600-future-manylinux-platform-tags-for-portable-linux-built-distributions/2414/8). 

As the original `libcrypt.so.1` was part of (g)libc, PyInstaller [explicitly excludes](https://github.com/pyinstaller/pyinstaller/blob/d3392e7597f69dbf60749f09b0ba4cdb7640089e/PyInstaller/depend/dylib.py#L201) it from it analysis (same as for other glibc libraries). Perhaps we should revise the rule so that only `libcrypt.so.1` is excluded, but we collect `libcrypt.so.2` - although I'd prefer if it was available on the target system (i.e., subject to system updates and patches) instead of application bundling its own copy.

---

FWIW, you could also build your python against `libcrypt.so.1` instead. I.e., if you look at 
```
sh-3.2# ls -l /usr/local/lib/libcrypt.*
-rw-r--r-- 1 root root 1144720 Feb  8 23:54 /usr/local/lib/libcrypt.a
-rwxr-xr-x 1 root root     923 Feb  8 23:54 /usr/local/lib/libcrypt.la
lrwxrwxrwx 1 root root      17 Feb  8 23:54 /usr/local/lib/libcrypt.so -> libcrypt.so.2.0.0
lrwxrwxrwx 1 root root      17 Feb  8 23:55 /usr/local/lib/libcrypt.so.1 -> libcrypt.so.1.1.0
-rwxr-xr-x 1 root root  765421 Feb  8 23:55 /usr/local/lib/libcrypt.so.1.1.0
lrwxrwxrwx 1 root root      17 Feb  8 23:54 /usr/local/lib/libcrypt.so.2 -> libcrypt.so.2.0.0
-rwxr-xr-x 1 root root  774288 Feb  8 23:54 /usr/local/lib/libcrypt.so.2.0.0
``` 
in your container, you'll see that `libcrypt.so` is symlinked to `libcrypt.so.2.0.0`; thus linking against `-lcrypt` ends up linking against `libcrypt.so.2`. If you replace that `libcrypt.so` with symlink to `libcrypt.so.1.1.0`, your python should end up linked against `libcrypt.so.1`.

Do note, though, that distributions that provide `libcrypt.so.2` as part of their system (such as Fedora) may not install compat `libcrypt.so.1` by default, so your application will end up with reverse problem unless user explicitly installs package that provides that library (in Fedora's case, `libxcrypt-compat`).

But this sort of mess is par for the course when dealing with binary-distributed applications across different linux distributions...

**Comment by dreambeyondorange** [comment]
Thanks for the feedback!  This lines up with what I was reading here on other tickets.  I think the tricky about excluding libcrypt.so.2 is that I don't see a way to install it with a package manager on Amazon linux or red hat (I haven't checked Ubuntu).  I will try symlinking to libcrypt.so.1.

After setting the libcrypt link to libcrypt.so.1, python3 is actually not linked to either one
```
[root@b9cf52c52c94 /]# ldd /usr/local/cpython-exe-builder/bin/python3
        linux-vdso.so.1 =>  (0x00007ffc3e1d1000)
        libpython3.9.so.1.0 => /usr/local/cpython-exe-builder/lib/libpython3.9.so.1.0 (0x00007f32cf57c000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f32cf360000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f32cf15c000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f32cef59000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f32ced50000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f32ceacd000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f32ce774000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f32cfb51000)
```

Which is what I expected from the build as I see it on other systems

**Comment by rokm** [comment]
What about the `_crypt` extension (`/usr/local/cpython-exe-builder/lib/python3.9/lib-dynload/_crypt.cpython-39-x86_64-linux-gnu.so`)? That should be linked against `libcrypt.so.1` now... and I think that's the part of python that actually uses `libcrypt`.

As per https://bugs.python.org/issue45433, the python shared library should not be linked against `libcrypt` in the first place. And I suspect the difference is that with `libcrypt.so` symlinked to `libcrypto.so.1.1.0`, the `crypt_r` symbol cannot be resolved by `configure` anymore:
```
...
checking for library containing crypt... no
checking for library containing crypt_r... no
checking for crypt_r... no
...
```
which prevents the python shared library from being linked against `-lcrypt`. I'm not sure how it impacts `_crypt` extension, so if you use the `crypt` module, maybe check if that still works.

FWIW, since python library is not really using `libcrypt` symbols, building with  `--as-needed `  linker option (`export LDFLAGS="${LDFLAGS} -Wl,--as-needed"` before building python) might also result in the library not actually ending up linked against `libcrypt.so.2.0.0` (but the `_crypt` extension should - which might or might pose problems with portability again, but only if you are actually using it).

**Comment by dreambeyondorange** [comment]
AFAIK we don't use the crypt extension, but I will try the flags instead of re-linking to see if I get the same result.  I don't see _crypt in the lib-dynload/ dir with the new symlink

After recompiling with the -Wl,--as-needed flags, I seem to get the same good result!
```
[root@6947c7516e80 /]# echo $LDFLAGS
-Wl,--as-needed

[root@6947c7516e80 /]# ll /usr/local/lib/libcrypt.so
lrwxrwxrwx 1 root root 17 Feb  8 23:54 /usr/local/lib/libcrypt.so -> libcrypt.so.2.0.0

[root@cb5a9f930254 /]# ldd /usr/local/cpython-exe-builder/bin/python3
        linux-vdso.so.1 =>  (0x00007ffc23772000)
        libpython3.9.so.1.0 => /usr/local/cpython-exe-builder/lib/libpython3.9.so.1.0 (0x00007f6c37680000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f6c37464000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f6c3710b000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f6c36f07000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f6c36d04000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f6c36afb000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f6c36878000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f6c37c55000)
```

**Comment by pkit** [comment]
For people getting here, just add to spec:

```python
a = Analysis([....],
    binaries=[('/usr/local/lib/libcrypt.so.2', '.')],
    ...,
)
```

**Comment by bwoodsend** [comment]
> For people getting here, just add to spec:
> 
> ```python
> a = Analysis([....],
>     binaries=[('/usr/local/lib/libcrypt.so.2', '.')],
>     ...,
> )
> ```

Vendoring `libcrypt` is very unhealthy for security.

**Comment by pkit** [comment]
@bwoodsend I suppose you haven't read the thread?
TL;DR libpython doesn't really use libcrypt, it's a bogus dependency that [will be completely removed in 3.11/3.12 ](https://bugs.python.org/issue45433)

**Comment by bwoodsend** [comment]
Yeah, you're right. Ignore me.

**Comment by dreambeyondorange** [comment]
Actually, I think the better approach is to use the `--as-needed` flag to prevent the library from being linked in the first place

**Comment by billyrudi** [comment]
can you write where i can put --as-needed? thanks

**Comment by dreambeyondorange** [comment]
> can you write where i can put --as-needed? thanks

I updated the LDFLAGS in my docker image when compiling Python

```
ENV LDFLAGS="${LDFLAGS} -Wl,--as-needed"
```
